### DISCUSSION TITLE
how to test the dataset with efficient_ad and torch_inference?

### DISCUSSION THREAD

**Discussion by Erwin2233** [open-edge-platform/anomalib#1283] [Q&A]
If I use torch_inference to load efficient_ad and it will report these.
>  Traceback (most recent call last):
> 
>   File "E:\anomalib-main\inference.py", line 117, in <module>
>     infer(weights="E:\\anomalib-main\\results\efficient_ad\Zn_eff_03_m\\run\weights\\torch\\model.pt")
>   File "E:\anomalib-main\inference.py", line 49, in infer
>     predictions = inferencer.predict(image=image)
>   File "E:\anomalib-main\src\anomalib\deploy\inferencers\base_inferencer.py", line 84, in predict
>     output = self.post_process(predictions, metadata=metadata)
>   File "E:\anomalib-main\src\anomalib\deploy\inferencers\torch_inferencer.py", line 139, in post_process
>     if isinstance(predictions[1], (Tensor)):
> KeyError: 1

**Comment by samet-akcay** [comment]
@YuRongHao2233 have you set the export mode to torch? By default anomalib saves the model based on Lightning. If you want to use torch inferencer you need to export the model as torch

**Reply by Erwin2233** [reply]
I have exported it as torch

**Reply by Erwin2233** [reply]
I have review the output of efficient_ad,it is a dic,so it can not be used as 

> if isinstance(predictions[1], (Tensor))

**Reply by samet-akcay** [reply]
Alright then, we'll have a look.

**Reply by blaz-r** [reply]
It seems like Efficient AD is returning dictionary, where other models return tensor or tuple. One thing I also noticed is that torch inferencer does have this [tuple check for patchcore](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/deploy/inferencers/torch_inferencer.py#L135-L145), while [openvino inferencer doesn't](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/deploy/inferencers/openvino_inferencer.py#L152-L158), so the score returned by patchcore isn't really accounted for, or am I missing something?

@YuRongHao2233 a quick fix in this case would be to just modify [Efficient AD torch model forward function](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/efficient_ad/torch_model.py#L351). I believe you can just return combined map in this case and it should work with openvino (this is probably not a good general solution).

**Comment by blaz-r** [comment]
@alexriedel1 how exactly are the results from forward pass used in case of inference? Are non-combined values even used, because [it seems they aren't.](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/efficient_ad/lightning_model.py#L265)

**Reply by alexriedel1** [reply]
yes thats right, only the combined feature map is used. you can also read it in the torch model https://github.com/openvinotoolkit/anomalib/blob/cdedf05451cc57fb373866595e8ef23a820cf123/src/anomalib/models/efficient_ad/torch_model.py#L302-L351
